<h1>ACTIVITY-1</h1>
<hr>
<p>Jefferson Emradora <p>

In [ ]:
#libraries

In [1]:
import numpy as np 
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
penguins = pd.read_csv('penguins.csv')

In [3]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h2>Descripte Analytics</h2>

In [4]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [5]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h1>Filtering a Column/Data</h1>

In [46]:
penguins['species'] = 'Adelie'

In [47]:
penguins['sex']

0        male
1      female
2      female
3         NaN
4      female
        ...  
339      male
340    female
341      male
342      male
343    female
Name: sex, Length: 344, dtype: category
Categories (2, object): ['female', 'male']

In [48]:
penguins[penguins['species'] == "Adelie"]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen        39.10000           18.7              181.0   
1    Adelie  Torgersen        39.50000           17.4              186.0   
2    Adelie  Torgersen        40.30000           18.0              195.0   
3    Adelie  Torgersen        43.92193            NaN                NaN   
4    Adelie  Torgersen        36.70000           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Adelie      Dream        55.80000           19.8              207.0   
340  Adelie      Dream        43.50000           18.1              202.0   
341  Adelie      Dream        49.60000           18.2              193.0   
342  Adelie      Dream        50.80000           19.0              210.0   
343  Adelie      Dream        50.20000           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

In [9]:
penguins['species'].value_counts()

species
Adelie|    344
Name: count, dtype: int64

In [10]:
penguins['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

In [11]:
penguins['species'] 

0      Adelie|
1      Adelie|
2      Adelie|
3      Adelie|
4      Adelie|
        ...   
339    Adelie|
340    Adelie|
341    Adelie|
342    Adelie|
343    Adelie|
Name: species, Length: 344, dtype: object

<h1>Adding two Datas</h1>

In [12]:
penguins[(penguins['species'] == 'Adelie') & (penguins['island'] == 'Dream')]

Empty DataFrame
Columns: [species, island, bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g, sex, year]
Index: []

<h1>Cleaning the data frame</h1>

In [13]:
penguins.isnull()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      False   False           False          False              False   
1      False   False           False          False              False   
2      False   False           False          False              False   
3      False   False            True           True               True   
4      False   False           False          False              False   
..       ...     ...             ...            ...                ...   
339    False   False           False          False              False   
340    False   False           False          False              False   
341    False   False           False          False              False   
342    False   False           False          False              False   
343    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 8 columns]

In [14]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>Filling Up NUll Values</h1>

In [51]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

<h1>Changing Data types</h1>

In [52]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    object  
 1   island             344 non-null    object  
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    int64   
dtypes: category(1), float64(4), int64(1), object(2)
memory usage: 19.4+ KB


In [53]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category')
penguins['sex'] = penguins['sex'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [54]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
count      344.000000     344.000000         344.000000   344.000000
mean        43.921930      17.151170         200.915205  4201.754386
std          5.443643       1.969027          14.020657   799.613058
min         32.100000      13.100000         172.000000  2700.000000
25%         39.275000      15.600000         190.000000  3550.000000
50%         44.250000      17.300000         197.000000  4050.000000
75%         48.500000      18.700000         213.000000  4750.000000
max         59.600000      21.500000         231.000000  6300.000000

<h1>DATA TYPES</h1>

In [23]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                333 non-null    category
 7   year               344 non-null    int64   
dtypes: category(2), float64(4), int64(1), object(1)
memory usage: 17.2+ KB


<h1>Exploratory Data Analysis (EDSA) </h1>


<h3>Which Specie of Peguins amongst the species is the haviest?</h3> <hr>

In [24]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

In [25]:
adelie_df = penguins[penguins['species'] == 'Adelie']

In [26]:
penguins['body_mass_g'].mean()

np.float64(4201.754385964912)

penguins[penguins['species'] == 'Adelie']['body_mass_g'].mean()

In [27]:
#import matpltlib to avoid error in plt
import matplotlib.pyplot as plt

<h1>number 1. What is the average body mass for male vs female penguins?</h1> <hr
                                                                               >

In [28]:
penguins.groupby("sex")['body_mass_g'].mean()


sex
female    3862.272727
male      4545.684524
Name: body_mass_g, dtype: float64

<h1>Insight</h1>
<p>Male penguins are generally heavier than female penguins, with an average body mass of about 4546 g compared to 3862 g for females. This suggests a clear sexual dimorphism where males are larger-bodied than females.</p>

<h1>number2 . How many penguins are there per species?</h1> <hr>


In [29]:
penguins['species'].value_counts()

species
Adelie|    344
Name: count, dtype: int64

<h1>Insight</h1>
<p>There are 344 Adelie penguins in the dataset, which makes them the most common species recorded. This could mean that Adelie penguins are more plentiful in the area studied, or that they were simply the species most often included in the data collection.</p>

<h1>number 3. which island has the largest average penguins body mass? </h1> <hr>

In [30]:
penguins.groupby("island")['body_mass_g'].mean()


island
Biscoe       4716.017964
Dream        3712.903226
Torgersen    3706.372549
Name: body_mass_g, dtype: float64

<h2>Insight: </h2>
<p>Penguins on Biscoe Island have the largest average body mass compared to those on Dream and Torgersen islands, indicating that the environment or food availability on Biscoe Island may favor larger-bodied penguins. Notably, Gentoo penguins, the heaviest species, are predominantly found there.</p>

<h1>number 4.Which species has the smallest average bill depth?</h1> <hr>

In [31]:
penguins.groupby("species")['bill_depth_mm'].mean()

species
Adelie|    17.15117
Name: bill_depth_mm, dtype: float64

<h1>Insight</h1>
<p>The average bill depth of Adelie penguins is about 17.15 mm, showing the typical size of their bills in the dataset. This gives an idea of their physical characteristics, which can be useful for comparing them with other penguin species.</p>

<h1>number 5. How Many Male Penguins Are There In Torgensen Island? </h1> <hr>

In [70]:
penguins['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

In [71]:
penguins[penguins['island'] == 'Torgersen']['sex'].value_counts()

sex
male      28
female    24
Name: count, dtype: int64

<h1>Insight</h1>
<p>There are 28 male penguins residing on Torgersen Island.</p>

<h1>number 6. Which island has the smallest average bill length?</h1><hr>

In [34]:
avg_bill_length_island = penguins.groupby("island")['bill_length_mm'].mean()

In [35]:
avg_bill_length_island

island
Biscoe       45.249535
Dream        44.167742
Torgersen    39.046576
Name: bill_length_mm, dtype: float64

<h1>Insights</h1>
<p>Torgersen Island penguins have the shortest average bill length (39.05 mm) among the three islands, hinting at possible adaptations to unique environmental conditions or food sources compared to Biscoe and Dream.</p>

<h1>number 7. On average, do penguins from Dream Island have longer flippers than those from Torgersen?</h1> <hr>

In [37]:
avg_flipper_islands = penguins.groupby("island")['flipper_length_mm'].mean()
subset = avg_flipper_islands.loc[["Dream", "Torgersen"]]

In [38]:
avg_flipper_islands

island
Biscoe       209.706587
Dream        193.072581
Torgersen    191.196078
Name: flipper_length_mm, dtype: float64

<h1>Insight</h1>
<p>On average, penguins from Dream Island have slightly longer flippers compared to those from Torgersen Island. This difference may be linked to the species distribution between the two islands.</p>

<h1>number 8. Is there a noticeable difference in bill depth between males and females?</h1> <hr>

In [40]:
penguins.groupby("sex")['bill_depth_mm'].mean()


sex
female    16.425455
male      17.891071
Name: bill_depth_mm, dtype: float64

<h1>Insight</h1>
<p>Male penguins have a slightly deeper (thicker) bill at about 17.9 mm, while female penguins average 16.4 mm. This shows that males generally have thicker beaks than females, but the difference is small.</p>

<h1>number 9. How has the average penguin body mass changed over the years?</h1><hr>

In [41]:
avg_mass_year = penguins.groupby("year")['body_mass_g'].mean()

In [42]:
avg_mass_year

year
2007    4124.541284
2008    4266.666667
2009    4210.294118
Name: body_mass_g, dtype: float64

<h1>Insight</h1>
<p>Penguin body mass increased from 2007 (4124.54 g) to 2008 (4276.67 g), then slightly declined in 2009 (4210.29 g), showing a small fluctuation over the three years.</p>

<h1>number 10. Which penguin species has the smallest variation in body mass? </h1> <hr>

In [44]:
body_mass_variation = penguins.groupby("species")['body_mass_g'].std()

In [53]:
body_mass_variation

species
Adelie|    801.954536
Name: body_mass_g, dtype: float64

<h1>Insight</h1>
<p>The penguin species have the small variation o body mass is the Adelie </p>

<h1>number 11. Is there a correlation between bill depth to penguins body mass?</h1> <hr>

In [72]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      344 non-null    float64 
 4   flipper_length_mm  344 non-null    float64 
 5   body_mass_g        344 non-null    float64 
 6   sex                344 non-null    category
 7   year               344 non-null    category
dtypes: category(4), float64(4)
memory usage: 12.7 KB


In [75]:
correlation = penguins['bill_depth_mm'].corr(penguins['body_mass_g'])

In [76]:
correlation

np.float64(-0.4719156211860664)

<h1>Insight</h1>
<p>There is no observable relationship between a penguin's bill depth and its body mass, suggesting that variations in bill depth do not predict or correlate with how heavy a penguin is.</p>

<h1>number 12. What is the max body mass of Adelie species during 2009</h1> <hr>

In [78]:
(penguins['species']=='Adelie') & (penguins['year'] == 2009)

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Length: 344, dtype: bool

In [79]:
penguins[(penguins['species']=='Adelie') & (penguins['year'] == 2007)]['body_mass_g'].max()

np.float64(6300.0)

<h1>Insight</h1>
<p>base on the statistics the max body mass of adelie during 2009 is around 6300.0</p>